pkg/sync/parallel_syncer.go

```diff
@@ -0,0 +1,1706 @@
+	lsc.pageToken = pageToken
+	return nil
+}
+
+// SetNeedsExpansion marks that grant expansion is needed
+func (lsc *LocalStateContext) SetNeedsExpansion() {
+	lsc.needsExpansion = true
+}
+
+// SetHasExternalResourcesGrants marks that external resources were found
+func (lsc *LocalStateContext) SetHasExternalResourcesGrants() {
+	lsc.hasExternalResources = true
+}
+
+// ShouldFetchRelatedResources returns whether related resources should be fetched
+func (lsc *LocalStateContext) ShouldFetchRelatedResources() bool {
+	return lsc.shouldFetchRelated
+}
+
+// ParallelSyncConfig holds configuration for parallel sync operations
+type ParallelSyncConfig struct {
+	// Number of worker goroutines to use for parallel processing
+	WorkerCount int
+	// Default bucket for resource types not explicitly configured
+	// If a resource type doesn't specify a sync_bucket, this default will be used
+	DefaultBucket string
+}
+
+// DefaultParallelSyncConfig returns a default configuration
+func DefaultParallelSyncConfig() *ParallelSyncConfig {
+	return &ParallelSyncConfig{
+		WorkerCount:   1,
+		DefaultBucket: "", // Empty string means each resource type gets its own bucket
+	}
+}
+
+// WithWorkerCount sets the number of worker goroutines
+func (c *ParallelSyncConfig) WithWorkerCount(count int) *ParallelSyncConfig {
+	if count > 0 {
+		c.WorkerCount = count
+	}
+	return c
+}
+
+// WithDefaultBucket sets the default bucket for resource types that don't specify a sync_bucket
+func (c *ParallelSyncConfig) WithDefaultBucket(bucket string) *ParallelSyncConfig {
+	c.DefaultBucket = bucket
+	return c
+}
+
+// task represents a unit of work for the parallel syncer
+type task struct {
+	Action       Action
+	ResourceID   string
+	Priority     int              // Higher priority tasks are processed first
+	ResourceType *v2.ResourceType // The resource type for this task
+}
+
+// taskQueue manages the distribution of tasks to workers using dynamic bucketing
+type taskQueue struct {
+	bucketQueues map[string]chan *task // Map of bucket name to task channel
+	config       *ParallelSyncConfig
+	mu           sync.RWMutex
+	closed       bool
+}
+
+// newTaskQueue creates a new task queue
+func newTaskQueue(config *ParallelSyncConfig) *taskQueue {
+	// Initialize with an empty map of bucket queues
+	// Buckets will be created dynamically as tasks are added
+	return &taskQueue{
+		bucketQueues: make(map[string]chan *task),
+		config:       config,
+	}
+}
+
+// AddTask adds a task to the appropriate queue
+func (q *taskQueue) AddTask(t *task) error {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	if q.closed {
+		return errors.New("task queue is closed")
+	}
+
+	// Determine the bucket for this task
+	bucket := q.getBucketForTask(t)
+
+	// Create the bucket queue if it doesn't exist
+	if _, exists := q.bucketQueues[bucket]; !exists {
+		queueSize := q.config.WorkerCount * 10
+		q.bucketQueues[bucket] = make(chan *task, queueSize)
```

Comment:
Is there recovery for a dropped task?  Some call sites are adding tasks in a loop over a page.  I'd expect to always hit this error and presumably fail the sync with the default worker count of 1.